File: activeMQ/cpp-client/sender_async_test.cpp
#include <activemq/library/ActiveMQCPP.h>
#include <activemq/core/ActiveMQConnectionFactory.h>
#include <cms/Connection.h>
#include <cms/Session.h>
#include <cms/BytesMessage.h>
#include <iostream>
#include <fstream>
#include <future>
#include <vector>
#include "../../utils/cpp/stats_collector.hpp"
#include "../../utils/cpp/test_data_loader.hpp"
#include "../../utils/cpp/message_helpers.hpp"

using namespace activemq::core;
using namespace cms;
using namespace std;
using json = nlohmann::json;
using messaging::MessageEnvelope;
using message_helpers::get_current_time_ms;

struct TaskResult {
    bool success;
    string message_id;
    long long duration;
    string error;
};

TaskResult send_message_task(Connection* connection, const json& item) {
    TaskResult res;
    res.success = false;
    res.message_id = message_helpers::extract_message_id(item);
    res.duration = 0;
    
    try {
        auto_ptr<Session> session(connection->createSession(Session::AUTO_ACKNOWLEDGE));
        auto_ptr<Destination> replyDest(session->createTemporaryQueue());
        auto_ptr<MessageProducer> producer(session->createProducer(NULL));
        auto_ptr<MessageConsumer> consumer(session->createConsumer(replyDest.get()));

        int target = item.value("target", 0);
        long long msg_start = get_current_time_ms();
        
        // Create and send message
        MessageEnvelope envelope = message_helpers::create_data_envelope(item);
        string body = message_helpers::serialize_envelope(envelope);
        
        auto_ptr<BytesMessage> message(session->createBytesMessage((unsigned char*)body.data(), body.size()));
        message->setCMSReplyTo(replyDest.get());
        message->setCMSCorrelationID("corr-cpp-async-" + res.message_id);
        
        string destName = "test_queue_" + to_string(target);
        auto_ptr<Destination> destination(session->createQueue(destName));
        producer->send(destination.get(), message.get());
        
        // Wait for reply
        auto_ptr<Message> reply(consumer->receive(100));  // 100ms timeout
        if (reply.get()) {
            const BytesMessage* bytesReply = dynamic_cast<const BytesMessage*>(reply.get());
            if (bytesReply && bytesReply->getCMSCorrelationID() == message->getCMSCorrelationID()) {
                unsigned char* buffer = new unsigned char[bytesReply->getBodyLength()];
                bytesReply->readBytes(buffer, bytesReply->getBodyLength());
                string response((char*)buffer, bytesReply->getBodyLength());
                delete[] buffer;
                
                MessageEnvelope resp_envelope;
                if (message_helpers::parse_envelope(response, resp_envelope) && 
                    message_helpers::is_valid_ack(resp_envelope, res.message_id)) {
                    res.duration = get_current_time_ms() - msg_start;
                    res.success = true;
                } else {
                    res.error = "Invalid ACK";
                }
            }
        } else {
            res.error = "Timeout";
        }
    } catch (CMSException& e) {
        res.error = e.getMessage();
    }
    
    return res;
}

int main() {
    activemq::library::ActiveMQCPP::initializeLibrary();

    try {
        auto test_data = test_data_loader::loadTestData();

        MessageStats stats;
        stats.set_metadata({
            {"service", "ActiveMQ"},
            {"language", "C++"},
            {"async", true}
        });
        long long start_time = get_current_time_ms();

        cout << " [x] Starting ASYNC transfer of " << test_data.size() << " messages..." << endl;

        auto_ptr<ConnectionFactory> factory(ConnectionFactory::createCMSConnectionFactory("tcp://localhost:61616"));
        auto_ptr<Connection> connection(factory->createConnection());
        connection->start();

        vector<future<TaskResult>> futures;
        for (auto& item : test_data) {
            futures.push_back(async(launch::async, send_message_task, connection.get(), item));
        }

        for (auto& fut : futures) {
            TaskResult res = fut.get();
            if (res.success) {
                stats.record_message(true, res.duration);
                cout << " [OK] Message " << res.message_id << " acknowledged" << endl;
            } else {
                stats.record_message(false);
                cout << " [FAILED] Message " << res.message_id << ": " << res.error << endl;
            }
        }

        connection->close();

        long long end_time = get_current_time_ms();
        stats.set_duration(start_time, end_time);
        
        json report = stats.get_stats();

        cout << "\nTest Results (ASYNC):" << endl;
        cout << "total_sent: " << stats.sent_count << endl;
        cout << "total_received: " << stats.received_count << endl;
        cout << "duration_ms: " << stats.get_duration_ms() << endl;

        ofstream rf("logs/report.txt", ios::app);
        if (rf.good()) {
            rf << report.dump() << endl;
            rf.close();
        }

    } catch (CMSException& e) {
        e.printStackTrace();
    }

    activemq::library::ActiveMQCPP::shutdownLibrary();
    return 0;
}

File: activeMQ/python-client/sender_async_test.py
#!/usr/bin/env python3
"""ActiveMQ Python Sender - Async"""
import sys
import json
import asyncio
import time
import stomp
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *
from test_data_loader import load_test_data
from stats_collector import MessageStats


class AsyncReplyListener(stomp.ConnectionListener):
    def __init__(self, loop):
        self.loop = loop
        self.futures = {} # Map correlation_id -> Future
        
    def on_message(self, frame):
        if 'correlation-id' in frame.headers:
            corr_id = frame.headers['correlation-id']
            if corr_id in self.futures:
                future = self.futures[corr_id]
                if not future.done():
                    # Handle body
                    body = frame.body
                    if isinstance(body, str):
                        body = body.encode('latin-1')  # latin-1 preserves bytes 0-255
                        
                    self.loop.call_soon_threadsafe(future.set_result, body)


async def send_message_task(conn, listener, item):
    """Send a single message asynchronously."""
    result = {'success': False, 'message_id': '', 'duration': 0, 'error': ''}
    
    try:
        message_id = extract_message_id(item)
        result['message_id'] = message_id
        target = item.get('target', 0)
        
        dest = f"/queue/test_queue_{target}"
        reply_dest = '/temp-queue/replies-async'
        
        msg_start = get_current_time_ms()
        
        # Create and send message
        envelope = create_data_envelope(item)
        body = serialize_envelope(envelope)
        
        corr_id = f"corr-async-{message_id}"
        
        # Create future for response
        future = asyncio.Future()
        listener.futures[corr_id] = future
        
        conn.send(body=body, destination=dest, headers={
            'reply-to': reply_dest,
            'correlation-id': corr_id,
            'content-type': 'application/octet-stream'
        })
        
        # Wait for reply with timeout
        try:
            resp_data = await asyncio.wait_for(future, timeout=0.1) # 100ms
            
            resp_envelope = parse_envelope(resp_data)
            if is_valid_ack(resp_envelope, message_id):
                result['duration'] = get_current_time_ms() - msg_start
                result['success'] = True
            else:
                result['error'] = 'Invalid ACK'
                
        except asyncio.TimeoutError:
            result['error'] = 'Timeout'
        finally:
            if corr_id in listener.futures:
                del listener.futures[corr_id]
            
    except Exception as e:
        result['error'] = str(e)
    
    return result


async def run():
    test_data = load_test_data()
    
    stats = MessageStats()
    stats.set_metadata({
        'service': 'ActiveMQ',
        'language': 'Python',
        'async': True
    })
    start_time = get_current_time_ms()
    
    print(f" [x] Starting ASYNC transfer of {len(test_data)} messages...")
    
    conn = stomp.Connection([('localhost', 61613)], auto_decode=False)
    loop = asyncio.get_event_loop()
    listener = AsyncReplyListener(loop)
    conn.set_listener('', listener)
    conn.connect('admin', 'admin', wait=True)
    
    reply_dest = '/temp-queue/replies-async'
    conn.subscribe(destination=reply_dest, id=1, ack='auto')
    
    # Process
    tasks = [send_message_task(conn, listener, item) for item in test_data]
    results = await asyncio.gather(*tasks)
    
    conn.disconnect()
    
    for result in results:
        if result['success']:
            stats.record_message(True, result['duration'])
            print(f" [OK] Message {result['message_id']} acknowledged")
        else:
            stats.record_message(False)
            print(f" [FAILED] Message {result['message_id']}: {result['error']}")
    
    end_time = get_current_time_ms()
    stats.set_duration(start_time, end_time)
    
    report = stats.get_stats()
    
    print("\nTest Results (ASYNC):")
    print(f"total_sent: {stats.sent_count}")
    print(f"total_received: {stats.received_count}")
    print(f"duration_ms: {stats.get_duration_ms()}")
    
    with open('logs/report.txt', 'a') as f:
        f.write(json.dumps(report) + '\n')


def main():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    loop.run_until_complete(run())
    loop.close()


if __name__ == "__main__":
    main()
